internals/plan/extensions_test.go

```diff
@@ -0,0 +1,687 @@
+					y-field:
+						y1:
+							override: replace
+							a: a
+							b: b
+				`,
+			},
+		},
+		error: "cannot validate plan section .* cannot find .* as required by .*",
+	}, {
+		summary: "Check empty section omits entry",
+		extensions: map[string]plan.LayerSectionExtension{
+			"x-field": &xExtension{},
+			"y-field": &yExtension{},
+		},
+		layers: []*inputLayer{
+			{
+				order: 1,
+				label: "layer-x",
+				yaml: `
+					summary: x
+					description: desc-x
+					x-field:
+				`,
+			},
+			{
+				order: 2,
+				label: "layer-y",
+				yaml: `
+					summary: y
+					description: desc-y
+					y-field:
+				`,
+			},
+		},
+		result: &planResult{
+			x: &xSection{},
+			y: &ySection{},
+		},
+		resultYaml: "{}\n",
+	}, {
+		summary: "Load file layers",
+		extensions: map[string]plan.LayerSectionExtension{
+			"x-field": &xExtension{},
+			"y-field": &yExtension{},
+		},
+		layers: []*inputLayer{
+			{
+				order: 1,
+				label: "layer-x",
+				yaml: `
+					summary: x
+					description: desc-x
+					x-field:
+						x1:
+							override: replace
+							a: a
+							b: b
+							y-field:
+							  - y1
+				`,
+			},
+			{
+				order: 2,
+				label: "layer-y",
+				yaml: `
+					summary: y
+					description: desc-y
+					y-field:
+						y1:
+							override: replace
+							a: a
+							b: b
+				`,
+			},
+		},
+		result: &planResult{
+			x: &xSection{
+				Entries: map[string]*X{
+					"x1": &X{
+						Name:     "x1",
+						Override: plan.ReplaceOverride,
+						A:        "a",
+						B:        "b",
+						Y: []string{
+							"y1",
+						},
+					},
+				},
+			},
+			y: &ySection{
+				Entries: map[string]*Y{
+					"y1": &Y{
+						Name:     "y1",
+						Override: plan.ReplaceOverride,
+						A:        "a",
+						B:        "b",
+					},
+				},
+			},
+		},
+		resultYaml: string(reindent(`
+			x-field:
+				x1:
+					override: replace
+					a: a
+					b: b
+					y-field:
+						- y1
+			y-field:
+				y1:
+					override: replace
+					a: a
+					b: b`)),
+	},
+}
+
+func (s *S) TestPlanExtensions(c *C) {
+nexttest:
+	for testIndex, testData := range extensionTests {
+		c.Logf("TestPlanExtensions :: %s (data index %v)", testData.summary, testIndex)
+
+		// Write layers to test directory.
+		layersDir := filepath.Join(c.MkDir(), "layers")
+		s.writeLayerFiles(c, layersDir, testData.layers)
+		var p *plan.Plan
+
+		// Register test extensions.
+		for field, extension := range testData.extensions {
+			err := func() (err error) {
+				defer func() {
+					if r := recover(); r != nil {
+						err = fmt.Errorf("%v", r)
+					}
+				}()
+				plan.RegisterExtension(field, extension)
+				return nil
+			}()
+			if err != nil {
+				c.Assert(err, ErrorMatches, testData.error)
+				continue nexttest
+			}
+		}
+
+		// Load the plan layer from disk (parse, combine and validate).
+		p, err := plan.ReadDir(layersDir)
+		if testData.error != "" || err != nil {
+			// Expected error.
+			c.Assert(err, ErrorMatches, testData.error)
```

Comment:
May use continue nexttest here too and deindent.  The assumption that termination code will happen anyway due to the else isn't sound because any Assert in the branch below will break the flow too.

Reply:
Done. Much better, thank you. Resolving.